readme.md
# jRumble v1.4
*by Jack Rugile - http://jackrugile.com*
*extended by Matt Surabian - http://mattsurabian.com*

jRumble is a [jQuery](http://jquery.com) plugin that rumbles, vibrates, shakes, and rotates any element you choose. It's great to use as a hover effect or a way to direct attention to an element.

- **Latest Version:** 1.4
- **Latest Release Date:** May 22, 2013
- **Original Release Date:** March 30, 2011
- **Compressed:** 1.47kb
- **Uncompressed:** 4.84kb

### Documentation, Demos, and Discussion

Documentation, demos, and discussion can be found at http://jackrugile.com/jrumble

### Basic Usage

```javascript
// Initialize jRumble on Selector
$('#rumble-element').jrumble();

// Start rumble on element
$('#rumble-element').trigger('startRumble');

// Stop rumble on element
$('#rumble-element').trigger('stopRumble');

//To have the rumbling occur for a fixed amount of time, initialize with the rumbleTime option
$('#rumble-element').jrumble({
    rumbleTime:2000
});

// Trigger generic rumble event (note: if rumbleTime is not set, the rumble event is identical to startRumble)
$('#rumble-element').trigger('rumble');
```

### Options

<table>
	<thead>
		<tr>
			<th>Option</th>
			<th>Default</th>
			<th>Description</th>
		</tr>
	</thead>
	<tbody>
		<tr>
			<td><strong>x</strong></td>
			<td><em>2</em></td>
			<td>Set the horizontal rumble range (pixels)</td>
		</tr>
		<tr>
			<td><strong>y</strong></td>
			<td><em>2</em></td>
			<td>Set the vertical rumble range (pixels)</td>
		</tr>
		<tr>
			<td><strong>rotation</strong></td>
			<td><em>1</em></td>
			<td>Set the rotation range (degrees)</td>
		</tr>
		<tr>
			<td><strong>speed</strong></td>
			<td><em>15</em></td>
			<td>Set the speed/frequency in milliseconds between rumble movements (lower number = faster)</td>
		</tr>					
		<tr>
			<td><strong>opacity</strong></td>
			<td><em>false</em></td>
			<td>Activate opacity flickering while rumbling</td>
		</tr>
		<tr>
			<td><strong>opacityMin</strong></td>
			<td><em>.5</em></td>
			<td>When the opacity option is set to true, this controls the minimum opacity while flickering</td>
		</tr>
		<tr>
            <td><strong>rumbleTime</strong></td>
            <td><em>null</em></td>
            <td>By initializing the jRumble object with rumbleTime, the generic 'rumble' event can be used to save front end code.  After rumbleTime has elapsed, the rumbling effect will stop.</td>
        </tr>
	</tbody>
</table>

### Known Issues

- For rumble elements that are position fixed/absolute, they should instead be wrapped in an element that is fixed/absolute
- Rotation does not work in Internet Explorer 8 and below

changelog.txt
v1.4 - May 22, 2013
Added rumble time option and generic rumble event to simplify usage. - MS
Integrated grunt for linting and building minified source. - MS

v1.3 - December 3, 2011
Internet Explorer filter typo fix ("apha" changed to "alpha")

v1.2 - October 23, 2011
Plugin reworked to provide greater flexibility and easier binding
Rumble is now activated with trigger('startRumble') and deactivated with trigger('stopRumble')
Added an opacity flicker option
IE speed 0 bug fixed
IE9+ rotation now supported
Removed rumbleEvent option
Removed fixed/absolute positioning options, jRumble element requires wrapped div that is positioned properly
Compressed code size went from 3.8kb to 1.6kb

v1.1 - March 30, 2011
Fixed error that occured each time jRumble was triggered - Thanks to tornography for finding this
Hover bug fixed with release of jQuery 1.5.2 (1.5.1 was leaving hover state in the "on" position)

v1.0 - March 30, 2011
jRumble Initial Release
	

jquery-jrumble.js
/*
jRumble v1.2 - http://jackrugile.com/jrumble
by Jack Rugile - http://jackrugile.com
Copyright 2011, Jack Rugile
MIT license - http://www.opensource.org/licenses/mit-license.php
*/

(function($){
	$.fn.jrumble = function(options){
		
		// JRUMBLE OPTIONS
		//---------------------------------
		var defaults = {
			rangeX: 2,
			rangeY: 2,
			rangeRot: 1,
			rumbleSpeed: 10,
			rumbleEvent: 'hover',
			posX: 'left',
			posY: 'top'
		};

		var opt = $.extend(defaults, options);

		return this.each(function(){
			
			// VARIABLE DECLARATION
			//---------------------------------
			$obj = $(this);			
			var rumbleInterval;
			var rumbleOff = true;	
			var rangeX = opt.rangeX;
			var rangeY = opt.rangeY;
			var rangeRot = opt.rangeRot;
			rangeX = rangeX*2;
			rangeY = rangeY*2;
			rangeRot = rangeRot*2;
			var rumbleSpeed = opt.rumbleSpeed;			
			var objPosition = $obj.css('position');
			var objXrel = opt.posX;
			var objYrel = opt.posY;
			var objXmove;
			var objYmove;
			var inlineChange;
			
			// SET POSITION RELATION IF CHANGED
			//---------------------------------
			if(objXrel === 'left'){
				objXmove = parseInt($obj.css('left'),10);
			}
			if(objXrel === 'right'){
				objXmove = parseInt($obj.css('right'),10);
			}
			if(objYrel === 'top'){
				objYmove = parseInt($obj.css('top'),10);
			}
			if(objYrel === 'bottom'){
				objYmove = parseInt($obj.css('bottom'),10);
			}
			
			// RUMBLER FUNCTION
			//---------------------------------			
			function rumbler(elem) {				
				var randBool = Math.random();
				var randX = Math.floor(Math.random() * (rangeX+1)) -rangeX/2;
				var randY = Math.floor(Math.random() * (rangeY+1)) -rangeY/2;
				var randRot = Math.floor(Math.random() * (rangeRot+1)) -rangeRot/2;	
				
				// IF INLINE, MAKE INLINE-BLOCK FOR ROTATION
				//---------------------------------
				if(elem.css('display') === 'inline'){
					inlineChange = true;
					elem.css('display', 'inline-block')
				}
			
				// ENSURE MOVEMENT
				//---------------------------------			
				if(randX === 0 && rangeX !== 0){
					if(randBool < .5){
						randX = 1;
					}
					else {
						randX = -1;
					}
				}
				
				if(randY === 0 && rangeY !== 0){
					if(randBool < .5){
						randY = 1;
					}
					else {
						randY = -1;
					}
				}
				
				// RUMBLE BASED ON POSITION
				//---------------------------------
				if(objPosition === 'absolute'){
					elem.css({'position':'absolute','-webkit-transform': 'rotate('+randRot+'deg)', '-moz-transform': 'rotate('+randRot+'deg)', '-o-transform': 'rotate('+randRot+'deg)', 'transform': 'rotate('+randRot+'deg)'});
					elem.css(objXrel, objXmove+randX+'px');
					elem.css(objYrel, objYmove+randY+'px');
				}
				if(objPosition === 'fixed'){
					elem.css({'position':'fixed','-webkit-transform': 'rotate('+randRot+'deg)', '-moz-transform': 'rotate('+randRot+'deg)', '-o-transform': 'rotate('+randRot+'deg)', 'transform': 'rotate('+randRot+'deg)'});
					elem.css(objXrel, objXmove+randX+'px');
					elem.css(objYrel, objYmove+randY+'px');
				}
				if(objPosition === 'static' || objPosition === 'relative'){
					elem.css({'position':'relative','-webkit-transform': 'rotate('+randRot+'deg)', '-moz-transform': 'rotate('+randRot+'deg)', '-o-transform': 'rotate('+randRot+'deg)', 'transform': 'rotate('+randRot+'deg)'});
					elem.css(objXrel, randX+'px');
					elem.css(objYrel, randY+'px');
				}
			} // End rumbler function
			
			// EVENT TYPES (rumbleEvent)
			//---------------------------------	
			var resetRumblerCSS = {'position':objPosition,'-webkit-transform': 'rotate(0deg)', '-moz-transform': 'rotate(0deg)', '-o-transform': 'rotate(0deg)', 'transform': 'rotate(0deg)'};
			
			if(opt.rumbleEvent === 'hover'){
				$obj.hover(
					function() {
						var rumblee = $(this);
						rumbleInterval = setInterval(function() { rumbler(rumblee); }, rumbleSpeed);
					},
					function() {
						var rumblee = $(this);
						clearInterval(rumbleInterval);
						rumblee.css(resetRumblerCSS);
						rumblee.css(objXrel, objXmove+'px');
						rumblee.css(objYrel, objYmove+'px');
						if(inlineChange === true){
							rumblee.css('display','inline');
						}
					}
				);
			}
			
			if(opt.rumbleEvent === 'click'){
				$obj.toggle(function(){
					var rumblee = $(this);
					rumbleInterval = setInterval(function() { rumbler(rumblee); }, rumbleSpeed);
				}, function(){
					var rumblee = $(this);
					clearInterval(rumbleInterval);
					rumblee.css(resetRumblerCSS);
					rumblee.css(objXrel, objXmove+'px');
					rumblee.css(objYrel, objYmove+'px');
					if(inlineChange === true){
						rumblee.css('display','inline');
					}
				});
			}
			
			if(opt.rumbleEvent === 'mousedown'){
				$obj.bind({
					mousedown: function(){
						var rumblee = $(this);
						rumbleInterval = setInterval(function() { rumbler(rumblee); }, rumbleSpeed);
					}, 
					mouseup: function(){
						var rumblee = $(this);
						clearInterval(rumbleInterval);
						rumblee.css(resetRumblerCSS);
						rumblee.css(objXrel, objXmove+'px');
						rumblee.css(objYrel, objYmove+'px');
						if(inlineChange === true){
							rumblee.css('display','inline');
						}
					},
					mouseout: function(){
						var rumblee = $(this);
						clearInterval(rumbleInterval);
						rumblee.css(resetRumblerCSS);
						rumblee.css(objXrel, objXmove+'px');
						rumblee.css(objYrel, objYmove+'px');
						if(inlineChange === true){
							rumblee.css('display','inline');
						}
					}
				});
			}
			
			if(opt.rumbleEvent === 'dblclick'){
				$obj.bind({
					dblclick: function(){
						var rumblee = $(this);
						rumbleOff = !rumbleOff;
						if(rumbleOff){
							// toggle rumble off
							clearInterval(rumbleInterval);
							rumblee.css(resetRumblerCSS);
							rumblee.css(objXrel, objXmove+'px');
							rumblee.css(objYrel, objYmove+'px');
							if(inlineChange === true){
								rumblee.css('display','inline');
							}
						} else {
							rumbleInterval = setInterval(function() { rumbler(rumblee); }, rumbleSpeed);
						}
						
					}
				})
			}
			
			if(opt.rumbleEvent === 'constant'){
				var rumblee = $(this);
				rumbleInterval = setInterval(function() { rumbler(rumblee); }, rumbleSpeed);
			}
			
		});
	}; 
})(jQuery);

jquery-jrumble-min.js
/* jRumble v1.2 - http://jackrugile.com/jrumble
by Jack Rugile - http://jackrugile.com
Copyright 2011, Jack Rugile
MIT license - http://www.opensource.org/licenses/mit-license.php
*/
(function(a){a.fn.jrumble=function(b){var d={rangeX:2,rangeY:2,rangeRot:1,rumbleSpeed:10,rumbleEvent:"hover",posX:"left",posY:"top"};var c=a.extend(d,b);return this.each(function(){$obj=a(this);var s;var f=true;var j=c.rangeX;var i=c.rangeY;var r=c.rangeRot;j=j*2;i=i*2;r=r*2;var l=c.rumbleSpeed;var k=$obj.css("position");var g=c.posX;var q=c.posY;var n;var m;var p;if(g==="left"){n=parseInt($obj.css("left"),10);}if(g==="right"){n=parseInt($obj.css("right"),10);}if(q==="top"){m=parseInt($obj.css("top"),10);}if(q==="bottom"){m=parseInt($obj.css("bottom"),10);}function e(u){var w=Math.random();var x=Math.floor(Math.random()*(j+1))-j/2;var v=Math.floor(Math.random()*(i+1))-i/2;var t=Math.floor(Math.random()*(r+1))-r/2;if(u.css("display")==="inline"){p=true;u.css("display","inline-block");}if(x===0&&j!==0){if(w<0.5){x=1;}else{x=-1;}}if(v===0&&i!==0){if(w<0.5){v=1;}else{v=-1;}}if(k==="absolute"){u.css({position:"absolute","-webkit-transform":"rotate("+t+"deg)","-moz-transform":"rotate("+t+"deg)","-o-transform":"rotate("+t+"deg)",transform:"rotate("+t+"deg)"});u.css(g,n+x+"px");u.css(q,m+v+"px");}if(k==="fixed"){u.css({position:"fixed","-webkit-transform":"rotate("+t+"deg)","-moz-transform":"rotate("+t+"deg)","-o-transform":"rotate("+t+"deg)",transform:"rotate("+t+"deg)"});u.css(g,n+x+"px");u.css(q,m+v+"px");}if(k==="static"||k==="relative"){u.css({position:"relative","-webkit-transform":"rotate("+t+"deg)","-moz-transform":"rotate("+t+"deg)","-o-transform":"rotate("+t+"deg)",transform:"rotate("+t+"deg)"});u.css(g,x+"px");u.css(q,v+"px");}}var h={position:k,"-webkit-transform":"rotate(0deg)","-moz-transform":"rotate(0deg)","-o-transform":"rotate(0deg)",transform:"rotate(0deg)"};if(c.rumbleEvent==="hover"){$obj.hover(function(){var t=a(this);s=setInterval(function(){e(t);},l);},function(){var t=a(this);clearInterval(s);t.css(h);t.css(g,n+"px");t.css(q,m+"px");if(p===true){t.css("display","inline");}});}if(c.rumbleEvent==="click"){$obj.toggle(function(){var t=a(this);s=setInterval(function(){e(t);},l);},function(){var t=a(this);clearInterval(s);t.css(h);t.css(g,n+"px");t.css(q,m+"px");if(p===true){t.css("display","inline");}});}if(c.rumbleEvent==="mousedown"){$obj.bind({mousedown:function(){var t=a(this);s=setInterval(function(){e(t);},l);},mouseup:function(){var t=a(this);clearInterval(s);t.css(h);t.css(g,n+"px");t.css(q,m+"px");if(p===true){t.css("display","inline");}},mouseout:function(){var t=a(this);clearInterval(s);t.css(h);t.css(g,n+"px");t.css(q,m+"px");if(p===true){t.css("display","inline");}}});}if(c.rumbleEvent==="dblclick"){$obj.bind({dblclick:function(){var t=a(this);f=!f;if(f){clearInterval(s);t.css(h);t.css(g,n+"px");t.css(q,m+"px");if(p===true){t.css("display","inline");}}else{s=setInterval(function(){e(t);},l);}}});}if(c.rumbleEvent==="constant"){var o=a(this);s=setInterval(function(){e(o);},l);}});};})(jQuery);

jquery.jrumble.js
/*
jRumble v1.4 - http://jackrugile.com/jrumble
by Jack Rugile - http://jackrugile.com
extended by Matt Surabian - http://mattsurabian.com

MIT License
-----------------------------------------------------------------------------
Copyright (c) 2012 Jack Rugile, http://jackrugile.com
Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

(function($){
	$.fn.jrumble = function(options){
		
		/*========================================================*/
		/* Options
		/*========================================================*/
		var defaults = {
			x: 2,
			y: 2,
			rotation: 1,
			speed: 15,
			opacity: false,
			opacityMin: 0.5,
            rumbleTime: null
		},
		opt = $.extend(defaults, options);	
				
		return this.each(function(){

			/*========================================================*/
			/* Variables
			/*========================================================*/
			var $this = $(this),				
				x = opt.x*2,
				y = opt.y*2,
				rot = opt.rotation*2,
				speed = (opt.speed === 0) ? 1 : opt.speed,			
				opac = opt.opacity,
				opacm = opt.opacityMin,
                rumbleTime = opt.rumbleTime,
				inline,
				interval;
			
			/*========================================================*/
			/* Rumble Function
			/*========================================================*/		
			var rumbler = function(){				
				var rx = Math.floor(Math.random() * (x+1)) -x/2,
					ry = Math.floor(Math.random() * (y+1)) -y/2,
					rrot = Math.floor(Math.random() * (rot+1)) -rot/2,
					ropac = opac ? Math.random() + opacm : 1;
					
				/*========================================================*/
				/* Ensure Movement From Original Position
				/*========================================================*/				
				rx = (rx === 0 && x !== 0) ? ((Math.random() < 0.5) ? 1 : -1) : rx;
				ry = (ry === 0 && y !== 0) ? ((Math.random() < 0.5) ? 1 : -1) : ry;
				
				/*========================================================*/
				/* Check Inline
				/*========================================================*/
				if($this.css('display') === 'inline'){
					inline = true;
					$this.css('display', 'inline-block');
				}
				
				/*========================================================*/
				/* Rumble Element
				/*========================================================*/			
				$this.css({
					'position':'relative',
					'left':rx+'px',
					'top':ry+'px',
					'-ms-filter':'progid:DXImageTransform.Microsoft.Alpha(Opacity='+ropac*100+')',
					'filter':'alpha(opacity='+ropac*100+')',					
					'-moz-opacity':ropac,					
					'-khtml-opacity':ropac,					
					'opacity':ropac,
					'-webkit-transform':'rotate('+rrot+'deg)', 
					'-moz-transform':'rotate('+rrot+'deg)', 
					'-ms-transform':'rotate('+rrot+'deg)',
					'-o-transform':'rotate('+rrot+'deg)', 
					'transform':'rotate('+rrot+'deg)'
				});
			};
			
			/*========================================================*/
			/* Rumble CSS Reset
			/*========================================================*/
			var reset = {
				'left':0,
				'top':0,
				'-ms-filter':'progid:DXImageTransform.Microsoft.Alpha(Opacity=100)',
				'filter':'alpha(opacity=100)',					
				'-moz-opacity':1,					
				'-khtml-opacity':1,					
				'opacity':1,
				'-webkit-transform':'rotate(0deg)',
				'-moz-transform':'rotate(0deg)',
				'-ms-transform':'rotate(0deg)',
				'-o-transform':'rotate(0deg)',
				'transform':'rotate(0deg)'
			};
			
			/*========================================================*/
			/* Rumble Start/Stop Trigger
			/*========================================================*/
			$this.bind({
				'startRumble': function(e){
					e.stopPropagation();
					clearInterval(interval);
					interval = setInterval(rumbler, speed);
				},
				'stopRumble': function(e){
					e.stopPropagation();
					clearInterval(interval);
					if(inline){
						$this.css('display', 'inline');
					}
					$this.css(reset);
				},
                'rumble':function(e){
                    $this.trigger('startRumble');
                    if(rumbleTime !== null){
                        setTimeout(function(){
                            $this.trigger('stopRumble');
                        },rumbleTime);
                    }
                }
			});		
			
		});// End return this.each
	};// End $.fn.jrumble
})(jQuery);